### PR TITLE
[Bugfix] Show proper error messages in exam management

### PR DESCRIPTION
**Pull request by jhainzlschmid**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

When actions in the student exam management (like generating student exams, missing student exams, assessing unsubmitted student exams, ...) fail the client does not show a proper error message, but instead only shows "Bad Request" or something similar. Especially in the case of student exam generation the user should get detailed information about the failure.

### Description

The previous error handling only defaulted to generic error handling, which simply shows a translated error message depending on the response status of the query. I added a method, which is always called when one of related queries fails. This new method shows a translated string specific to the failed action and gives some reasoning (if provided in the request response). Otherwise it falls back to the original generic error handling.

### Steps for Testing

1. Create a new exam and give it a wrong configuration (e.g. set number of exercise groups to 2, but only add 1 exercise group)
2. Add some students to the exam (Exam Management -> Students -> Import students or Register course students)
2. Navigate to Exam Management -> Student Exams
3. Click on generate individual exams
4. Assert that the alert shows the specific reasoning why the action failed
5. Also try to cause errors in the other actions on this site (generate missing individual exams, prepare exercise start, unlock all repositories, lock all repositories, evaluate quizzes, assess unsubmitted student exams) and verify the alert messages

Note:
Other actions (except for generating individual exams or generating missing individual exams) do not provide that specific reasoning, so in this case only a translated message specific to that action is shown with the error.message appended if available. Otherwise the old error handling logic applies (this happens for example if the user has no internet connection and the request fails).

### Test Coverage

student-exams.component.ts 96%

### Screenshots

**Before**
image

**After**
image

### CODE REVIEW

**Review by lschlesinger — DISMISSED** [review]
I think this really improves UX for instructors. During testing on TS1, I produced following errors:
- Too few exercise groups
- No exercise in exercise group
- Wrong amount of exercise points

In every case I received the detailed error alert.

File: src/main/webapp/app/exam/manage/student-exams/student-exams.component.ts [hunk 1]
```diff
@@ -360,4 +360,22 @@ export class StudentExamsComponent implements OnInit {
...
+    /**
+     * Shows the translated error message if an error key is available in the error response. Otherwise it defaults to the generic alert.
+     * @param translationString the string identifier in the translation service for the text. This is ignored if the response does not contain an error message or error key.
+     * @param err the error response
+     */
+    private handleError(translationString: string, err: HttpErrorResponse) {
+        let errorDetail = err?.error?.message;
+        if (err?.error && err.error.errorKey) {
+            errorDetail = this.translateService.instant(err.error.errorKey);
+        }
```
Comment (FuchsDominik) [inline src/main/webapp/app/exam/manage/student-exams/student-exams.component.ts:373]:
```suggestion
        let errorDetail;
        if (err?.error && err.error.errorKey) {
            errorDetail = this.translateService.instant(err.error.errorKey);
        } else {
            errorDetail = err?.error?.message;
        }
```
Don't know how much of an improvement this is to you, but as if the errorKey exists, we overwrite the errorDetail anyway, we could throw this into an else statement to one assignment

**Review by FuchsDominik — APPROVED** [review]
Code looks very good :) Found one improvement which you could consider

File: src/main/webapp/app/exam/manage/student-exams/student-exams.component.ts (hunk 1 above)
Comment (jhainzlschmid) [reply src/main/webapp/app/exam/manage/student-exams/student-exams.component.ts:373]: Yeah I agree. Performance-wise it won't make much difference, but I think your suggestion is more readable. I will change it

**Review by FrankeLukas — APPROVED** [review]
Lgtm. I just noticed there is a space in front of the second row in that error message:
grafik
Not sure if that is intentional. (It doesn't look that bad imo.)

**Comment by jhainzlschmid** [comment]
> Lgtm. I just noticed there is a space in front of the second row in that error message:
> grafik
> Not sure if that is intentional. (It doesn't look that bad imo.)

Yeah I was also wondering about that. I used the already existing translation strings so the space was already there. I also thought it wasn't that bad, so I left it like that. If anyone has some hard feelings on the matter I will change it.

**Review by willergeorg — APPROVED** [review]
Tested on TS1. Works for me.

**Pull request merged**